### PR TITLE
 OZ-560: Rename project,Added E2E tests and registration Opt In

### PR DESCRIPTION
**Pull request by enyachoke**
This is a PR for the work in https://mekomsolutions.atlassian.net/browse/OZ-560. It introduces a number of changes to the original project;

- Container registration is now opt-in
- Upgrade a number of libraries
- Adds E2E Tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,164 @@
...
+## Running Container-to-Consul
+
+### Using Docker
+
+```
+docker run -d \
+  -e c2c_consul__host=localhost \
+  -e c2c_consul__port=8500 \
+  -v /var/run/docker.sock:/var/run/docker.sock \
+  mekomsolutions/container2sul
```
Comment (rbuisson) [inline README.md]: Should we keep this Docker image name?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,164 @@
-# container-to-consul
-A tool for automatic discovery and registration of Docker services to Consul using container labels.
+# Container-to-Consul
+
+[![Build Status](https://travis-ci.org/mekomsolutions/container-to-consul.svg?branch=master)](https://travis-ci.org/mekomsolutions/container-to-consul) 
+[![codecov](https://codecov.io/gh/mekomsolutions/container-to-consul/branch/master/graph/badge.svg)](https://codecov.io/gh/mekomsolutions/container-to-consul)
```
Comment (rbuisson) [inline README.md]: Are these going to work?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,164 @@
...
+### IP Address
+
+By default, the registered ip address is the first one found in the `NetworkSettings` group from `docker inspect` result.
+
+If the container is running in `host` network mode, the ip address used is the `docker0` one.
+
+The registered ip address can be configured using a `consul.ip` label.
+
+```
+docker run --rm --name containertoConsul -ti -l consul.ip=10.0.0.1 alpine ash
```
Comment (rbuisson) [inline README.md]: Let's have the container named `container-to-consul` if possible.

File: e2e/container2consul.spec.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import { test, expect } from '@playwright/test';
+import { startDockerCompose, stopDockerCompose } from './util';
+
+test.beforeAll(async () => {
+  await stopDockerCompose();
+  await startDockerCompose();
+  // Pause for 5 seconds to allow nginx test containers to start
+  await new Promise(r => setTimeout(r, 3000));
```
Comment (kdaud) [inline e2e/container2consul.spec.ts]:
```suggestion
  await new Promise(r => setTimeout(r, 5000));
```

File: README.md [hunk 5]
```diff
@@ -1,2 +1,164 @@
-# container-to-consul
-A tool for automatic discovery and registration of Docker services to Consul using container labels.
+# Container-to-Consul
+
+[![Build Status](https://travis-ci.org/mekomsolutions/container-to-consul.svg?branch=master)](https://travis-ci.org/mekomsolutions/container-to-consul) 
+[![codecov](https://codecov.io/gh/mekomsolutions/container-to-consul/branch/master/graph/badge.svg)](https://codecov.io/gh/mekomsolutions/container-to-consul)
+
+Container-to-Consul automatically registers and deregisters nodes for any Docker container.
+
+This project is inspired by [gliderlabs registrator](https://github.com/gliderlabs/registrator).
```
Comment (rbuisson) [inline README.md]: Isn't this based on https://github.com/benoitvidis/con-tainer2sul instead?

File: README.md [hunk 6]
```diff
@@ -1,2 +1,164 @@
-# container-to-consul
-A tool for automatic discovery and registration of Docker services to Consul using container labels.
+# Container-to-Consul
+
+[![Build Status](https://travis-ci.org/mekomsolutions/container-to-consul.svg?branch=master)](https://travis-ci.org/mekomsolutions/container-to-consul) 
+[![codecov](https://codecov.io/gh/mekomsolutions/container-to-consul/branch/master/graph/badge.svg)](https://codecov.io/gh/mekomsolutions/container-to-consul)
+
+Container-to-Consul automatically registers and deregisters nodes for any Docker container.
```
Comment (rbuisson) [inline README.md]:
```suggestion
Container-to-Consul automatically registers and deregisters any Docker container as HashiCorp Consul nodes.
```

**Review by kdaud — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (rbuisson) [reply README.md]: If so, it needs more emphasis.
What about:
> This project is largely based on the original work done in [con-tainer2sul](https://github.com/benoitvidis/con-tainer2sul) by [Benoît Vidis](https://github.com/benoitvidis/). While this fork includes significant changes and improvements, we would like to acknowledge the foundation provided by the original project, without which this version would not have been possible.

> Please note that some files, variables, or other elements might still reference the original project name. These remnants will be updated over time as the project evolves, but their presence reflects the project's roots as `con-tainer2sul`.

File: README.md
Comment (rbuisson) [inline README.md:1]: @enyachoke  the doc specifies multiple commands to be run, with different option. Can those options be used simultaneously? eg, using `consul.service=myservice` and `consul.tags=xyz`

**Comment by rbuisson** [comment]
@enyachoke  could you provide some instructions to test?
The Docker image `mekomsolutions/container2sul` does not seem to be available so how do I build it?

**Comment by enyachoke** [comment]
> @enyachoke could you provide some instructions to test? The Docker image `mekomsolutions/container2sul` does not seem to be available so how do I build it?

@rbuisson This will be available after we merge see
https://github.com/mekomsolutions/container-to-consul/blob/52f007b0788d2fecc29abd6c7b6d7fe8cce04fe6/.github/workflows/build.yml#L13-L19

File: e2e/container2consul.spec.ts (hunk 4 above)
Comment (enyachoke) [reply e2e/container2consul.spec.ts]: @kdaud  is this based on your tests locally?

File: README.md
Comment (enyachoke) [reply README.md:1]: Yes they can as used in the tests https://github.com/mekomsolutions/container-to-consul/blob/52f007b0788d2fecc29abd6c7b6d7fe8cce04fe6/e2e/docker/test-compose.yaml#L4-L10

File: e2e/container2consul.spec.ts (hunk 4 above)
Comment (kdaud) [reply e2e/container2consul.spec.ts]: Setting 3000 milliseconds equates to 3 seconds, and the comment indicates that a pause of 5 minutes is required to allow the Nginx test containers to start.

**Comment by rbuisson** [comment]
Yes I know. So what are the instructions?
docker build first? (what exact command). And then what?

File: README.md
Comment (rbuisson) [reply README.md:1]: In that case, the structure of the doc is misleading.
Why would someone want to run those commands one by one like that?

I suppose it is better to simply document what the options can do, without an example `docker run --rm ...` which in fact will probably never be used.

Comment (enyachoke) [reply README.md:1]: > In that case, the structure of the doc is misleading. Why would someone want to run those commands one by one like that?
> 
> I suppose it is better to simply document what the options can do, without an example `docker run --rm ...` which in fact will probably never be used.

The commands are used to explain the effect of the option and are not necessarily complete usage examples.

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+on:
+  push:
+    branches: [main]
+  pull_request:
+    branches: [main]
+  release:
+    types: [published]
+
+jobs:
+  docker-release:
```
Comment (rbuisson) [inline .github/workflows/build.yml]:
```suggestion
  docker-publish:
```

File: package.json [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  "version": "0.1.0",
+  "description": "Auto-register Docker containers in Consul",
+  "main": "lib/index.js",
+  "scripts": {
+    "test": "npm run --silent lint && npm run --silent unit-tests",
+    "lint": "eslint -c eslint.config.mjs lib",
+    "unit-tests": "istanbul cover _mocha -- --recursive --reporter spec --require should --require should-sinon"
+  },
+  "author": "Benoît Vidis",
+  "license": "Apache-2.0",
```
Comment (rbuisson) [inline package.json:13]: Could you add a new key for contributors:

`
  "contributors": [
    "Mekom Solutions (https://mekomsolutions.com/)"
  ]
`

File: README.md (hunk 2 above)
Comment (rbuisson) [reply README.md]: Did you remove these lines?

File: README.md [hunk 9]
```diff
@@ -1,2 +1,145 @@
...
+| `consul.port` | The port that the service will be discovered on| `none` |
+| `consul.service` | The name of the service in Consul | `container name` |
+| `consul.tags` | The tags that will be added to the service in Consul | `[]` |
+| `consul.kv.<KEY>` | Key/Value pairs | `{}` |
+
+### Registration
+Running;
+
+```
+docker run --name docker-nginx -d -l consul.register=true -l consul.ip=10.0.0.1 -l consul.service=nginx-service -l consul.tags=www,api -l consul.port=80 -l consul.kv.foo=bar -l consul.kv.bar=baz  nginx
```
Comment (rbuisson) [inline README.md]:
```suggestion
docker run --name nginx -d \
 -l consul.register=true \
  -l consul.ip=10.0.0.1 \
   -l consul.service=nginx \
    -l consul.tags=www,api \
     -l consul.port=80 \
      -l consul.kv.foo=bar \
       -l consul.kv.bar=baz \
        nginx
```

File: README.md [hunk 10]
```diff
@@ -1,2 +1,145 @@
...
+- port `80`
+- 2 key/value pairs under the `/services/nginx-service/` tree.
+
+You can check for the registered service in [Consul](http://localhost:8500/ui/dc1/services)
+
+> For a complete usage example please [see](https://github.com/mekomsolutions/container-to-consul/blob/main/e2e/docker/docker-compose.yaml).
+
+### Deregistration
+
+To deregister the container, you can simply stop it and remove it.
```
Comment (rbuisson) [inline README.md]: Is removing the container necessary to deregister? I don't think so, right?
I suggest to then remove the `rm` command to put it a bit after, as a note or something.

**Review by rbuisson — CHANGES_REQUESTED** [review]
I've pushed a change on the README to add some screenshots.
I've left some comments as well.

But it works!
Thanks, that's an awesome tool!

**Review by kdaud — APPROVED** [review]
LGTM

@rbuisson needs to do a final review of the PR.

**Review by rbuisson — APPROVED** [review]

**Pull request merged**